### PR TITLE
Add DirectedCodegen to make codegen simpler

### PR DESCRIPTION
**Pull request by mtdowling**
DirectedCodegen is an attempt to make using all the different pieces of
smithy-codegen-core easier. It wires integrations together with your
generator, automatically calls all the methods of integrations in the
appropriate order, executes the methods of the directed code generator,
and provides well-defined methods that help guide new code generation
projects. The goal is to point people in the right direction using
methods rather than just provide a blank slate and expect everything to
built from the ground up.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    /**
+     * Sets the required class used for SmithyIntegrations.
+     *
+     * @param integrationClass SmithyIntegration class.
+     */
+    public void integrationClass(Class<I> integrationClass) {
+        this.integrationClass = integrationClass;
+    }
```
Comment (JordonPhillips) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java:107]: Why is this needed?

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+    private void generateUnionShapes(C context, ServiceShape serviceShape, Set<Shape> shapes) {
+        for (UnionShape shape : model.getUnionShapes()) {
+            if (shapes.contains(shape)) {
+                LOGGER.finest(() -> "Generating union " + shape.getId());
+                directedCodegen.generateUnion(new GenerateUnion<>(context, serviceShape, shape));
+            }
+        }
+    }
+
+    private void generateEnumShapes(C context, ServiceShape serviceShape, Set<Shape> shapes) {
```
Comment (JordonPhillips) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java:405]: So current generators will have to use postProcess to generate enums right?

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/GenerateResource.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import software.amazon.smithy.model.shapes.ServiceShape;
+
+/**
+ * Directive used to generate a resource.
+ *
+ * @param <C> CodegenContext type.
+ * @param <S> Codegen settings type.
+ * @see DirectedCodegen#generateResource
+ */
+public final class GenerateResource<C extends CodegenContext<S, ?>, S> extends ShapeDirective<ResourceShape, C, S> {
```
Comment (JordonPhillips) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/GenerateResource.java:29]: nit: these all have action names, but they don't actually perform actions, they just hold arguments used to perform those actions

**Review by JordonPhillips — CHANGES_REQUESTED** [review]

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java (hunk 1 above)
Comment (mtdowling) [reply smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java:107]: To find them via SPI

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/GenerateResource.java (hunk 3 above)
Comment (mtdowling) [reply smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/GenerateResource.java:29]: They are tied 1:1 with the methods invoked on DirectedCodegen. I considered putting "Context" on the end or something, but with the generics, it makes the methods _really_ long.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java (hunk 2 above)
Comment (mtdowling) [reply smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java:405]: String enums? Yes, or they can do it in generateService.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+     * @return Returns the updated model.
+     */
+    public static Model simplifyModelForServiceCodegen(Model model, ShapeId service, ModelTransformer transformer) {
+        ServiceShape serviceShape = model.expectShape(service, ServiceShape.class);
+        model = transformer.copyServiceErrorsToOperations(model, serviceShape);
+        // model = transformer.flattenAndRemoveMixins(model);
+        model = transformer.copyServiceErrorsToOperations(model, serviceShape);
+        // model = transformer.changeStringEnumsToEnumShapes(model, true);
+        return model;
+    }
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java:98]: Yes

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/Directive.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.neighbor.Walker;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+
+/**
+ * Base class for all Directives emitted from {@link DirectedCodegen}.
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/Directive.java]: Done

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java (hunk 2 above)
Comment (mtdowling) [reply smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java:407]: I pushed a commit to add a directive for generating enum strings, and it can be extended in the future for generating enum shapes.

**Review by kstich — APPROVED** [review]

**Review by kstich — APPROVED** [review]

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/GenerateResource.java (hunk 3 above)
Comment (mtdowling) [reply smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/GenerateResource.java]: I went ahead and added a Directive suffix to these classes.

**Review by JordonPhillips — APPROVED** [review]
